clippy_lints/src/methods/collapsible_str_replace.rs

```diff
@@ -0,0 +1,275 @@
+            }
+        }
+    }
+}
+
+/// Check a chain of `str::replace` calls for `collapsible_str_replace` lint.
+fn check_consecutive_replace_calls<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>) {
+    if_chain! {
+        if let Some(from_args) = get_replace_call_from_args_if_all_char_ty(cx, expr);
+        if let Some(to_arg) = get_replace_call_unique_to_arg_repr(expr);
+        then {
+            let earliest_replace_call_span = get_earliest_replace_call_span(expr);
+
+            if replace_call_from_args_are_only_lit_chars(&from_args) {
+                let from_arg_reprs: Vec<String> = from_args.iter().map(|from_arg| {
+                    get_replace_call_char_arg_repr(from_arg).unwrap()
+                }).collect();
+                let app = Applicability::MachineApplicable;
+
+                span_lint_and_sugg(
+                    cx,
+                    COLLAPSIBLE_STR_REPLACE,
+                    expr.span.with_lo(earliest_replace_call_span.lo()),
+                    "used consecutive `str::replace` call",
+                    "replace with",
+                    format!(
+                        "replace(|c| matches!(c, {}), {})",
+                        from_arg_reprs.join(" | "),
+                        to_arg,
+                    ),
+                    app,
+                );
+            } else {
+                // Use fallback lint
+                let from_arg_reprs: Vec<String> = from_args.iter().map(|from_arg| {
+                    get_replace_call_char_arg_repr(from_arg).unwrap()
+                }).collect();
+                let app = Applicability::MachineApplicable;
+
+                span_lint_and_sugg(
+                    cx,
+                    COLLAPSIBLE_STR_REPLACE,
+                    expr.span.with_lo(earliest_replace_call_span.lo()),
+                    "used consecutive `str::replace` call",
+                    "replace with",
+                    format!(
+                        "replace(&[{}], {})",
+                        from_arg_reprs.join(" , "),
+                        to_arg,
+                    ),
+                    app,
+                );
+            }
+        }
+    }
+}
+
+/// Check if all the `from` arguments of a chain of consecutive calls to `str::replace`
+/// are all of `ExprKind::Lit` types. If any is not, return false.
+fn replace_call_from_args_are_only_lit_chars<'tcx>(from_args: &[&'tcx hir::Expr<'tcx>]) -> bool {
+    let mut only_lit_chars = true;
+
+    for from_arg in from_args.iter() {
+        match from_arg.kind {
+            ExprKind::Lit(..) => {},
+            _ => only_lit_chars = false,
+        }
+    }
+
+    only_lit_chars
+}
+
+/// Collect and return all of the `from` arguments of a chain of consecutive `str::replace` calls
+/// if these `from` arguments's expressions are of the `ty::Char` kind. Otherwise return `None`.
+fn get_replace_call_from_args_if_all_char_ty<'tcx>(
```

Comment:
Function names like this scream that this function does too much and is too complex.

Why not have one function that simply walks up the method call chain (so the `if let Some(("replace", [_, args @ ..], _)) = method_call(e)` part) and collects all the `replace` `MethodCall`s and then one function that takes those exprs and only has to collect the args. 

In the second function, if you find a argument that is **not** a `char`, just early return `None`.

Reply:
On second thought, maybe you want to do this while collecting the `MethodCall` expressions. 

But the method should have a signature like:

```rust
fn collect_replace_calls(..., to_arg: &Expr) -> ReplaceMethods
```

where 

```rust
struct ReplaceMethods {
    methods: Vec<&Expr>,
    args: Vec<&Expr>,
}
```

and in that function you should collect as long as (pseudo code)

```rust
SpanlessEq(to_arg, current_to_arg) && expr_ty(arg).peel_refs().is_char()
```